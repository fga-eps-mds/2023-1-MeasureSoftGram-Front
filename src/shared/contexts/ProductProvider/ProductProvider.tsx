import React, { createContext, useState, useContext, ReactNode, useMemo, useCallback, useEffect } from 'react';

import { Product } from '@customTypes/product';
import { productQuery } from '@services/product';
import { useOrganizationContext } from '@contexts/OrganizationProvider';

interface Props {
  children: ReactNode;
}

interface IProductContext {
  currentProduct?: Product;
  setCurrentProduct: (product: Product) => void;
  productsList?: Product[];
  updateProductList: (products: Product[]) => void;
}

const ProductContext = createContext<IProductContext | undefined>(undefined);

export function ProductProvider({ children }: Props) {
  const [currentProduct, setCurrentProduct] = useState<Product | undefined>();
  const [productsList, setProductsList] = useState<Product[]>([]);

  const { currentOrganization } = useOrganizationContext();

  const updateProductList = useCallback((products: Product[]) => {
    setProductsList(products);
  }, []);

  const loadAllProducts = async () => {
    try {
      const result = await productQuery.getAllProducts(currentOrganization.id);

      updateProductList(result.data.results);
    } catch (error) {
      console.error(error);
    }
  };

  useEffect(() => {
    if (currentOrganization) {
      loadAllProducts();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [currentOrganization]);

  const value = useMemo(
    () => ({
      currentProduct,
      setCurrentProduct,
      productsList,
      updateProductList
    }),
    [currentProduct, productsList, updateProductList]
  );

  return <ProductContext.Provider value={value}>{children}</ProductContext.Provider>;
}

export function useProductContext() {
  const context = useContext(ProductContext);

  if (context === undefined) {
    throw new Error('useProductContext must be used within a ProductContext');
  }

  return context;
}
