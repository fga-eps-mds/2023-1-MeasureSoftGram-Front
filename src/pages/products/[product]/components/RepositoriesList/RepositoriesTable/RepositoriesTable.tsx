import React, { useState, useEffect } from 'react';
import { toast } from 'react-toastify';
import { useRouter } from 'next/router';

import {
  TableContainer,
  Table,
  TableCell,
  TableHead,
  TableRow,
  TableBody,
  Collapse,
  Box,
  Typography
} from '@mui/material';
import { ArrowCircleRight, AddCircle, RemoveCircle } from '@mui/icons-material';

import { repository } from '@services/repository';
import { useRepositoryContext } from '@contexts/RepositoryProvider';
import { useProductContext } from '@contexts/ProductProvider';
import { useOrganizationContext } from '@contexts/OrganizationProvider';
import MeasureSoftGramChart from '@components/MeasureSoftGramChart';
import { HistoricalCharacteristicsProps, Historical } from '@customTypes/repository';

function RepositoriesTable() {
  const { currentProduct } = useProductContext();
  const { currentOrganization } = useOrganizationContext();
  const { repositoryList } = useRepositoryContext();
  const router = useRouter();

  const [open, setOpen] = useState({});

  const handleClickRedirects = (id: string) => {
    const path = `${currentOrganization?.id}-${currentProduct?.id}-${currentProduct?.name}/${id}`;
    router.push(path);
  };

  const [historicalCharacteristics, setHistoricalCharacteristics] = useState<Historical[]>([]);

  const handleClickExpand = (id: string) => {
    loaldHistoricalCharacteristics(id);
    setOpen((prevState) => ({
      ...prevState,
      [id]: !prevState[id]
    }));
  };

  const handleClickCollapse = (id: string) => {
    setOpen((prevState) => ({
      ...prevState,
      [id]: false
    }));
  };

  const loaldHistoricalCharacteristics = async (id: string) => {
    try {
      const organizationId = currentOrganization?.id;
      const productId = currentProduct?.id;
      const entity = 'characteristics';
      const repositoryId = id;

      const props: HistoricalCharacteristicsProps = {
        organizationId,
        productId,
        repositoryId,
        entity
      };

      const response = await repository.getHistorical(props);

      if (response?.data?.length) {
        return {
          setHistoricalCharacteristics: {}
        };
      }
      setHistoricalCharacteristics(response?.data.results);
    } catch (err) {
      toast.error(err);
    }
  };

  // update historicalCharacteristics if repositoryList changes
  useEffect(() => {
    if (repositoryList?.length) {
      loaldHistoricalCharacteristics(repositoryList[0].id);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [repositoryList]);

  return (
    <TableContainer>
      <Table>
        <TableHead>
          <TableRow>
            <TableCell>Nome</TableCell>
            <TableCell />
          </TableRow>
        </TableHead>
        <TableBody>
          {repositoryList?.map((repo) => (
            <React.Fragment key={repo.id}>
              <TableRow hover style={{ cursor: 'pointer' }} data-testid="repository-row">
                <TableCell>{repo.name}</TableCell>
                <TableCell align="right">
                  <ArrowCircleRight
                    aria-label="expand row"
                    onClick={() => handleClickRedirects(`${repo.id}-${repo.name}`)}
                  />
                  {open[repo.id] ? (
                    <RemoveCircle aria-label="collapse row" onClick={() => handleClickCollapse(repo.id)} />
                  ) : (
                    <AddCircle aria-label="expand row" onClick={() => handleClickExpand(repo.id)} />
                  )}
                </TableCell>
              </TableRow>
              <TableRow>
                <TableCell style={{ paddingBottom: 0, paddingTop: 0 }} colSpan={6}>
                  <Collapse in={open[repo.id]} timeout="auto" unmountOnExit>
                    <Box sx={{ margin: 1 }}>
                      {historicalCharacteristics?.length === 0 ? (
                        <Typography>Não é possível visualizar o gráfico, pois não há dados.</Typography>
                      ) : (
                        <MeasureSoftGramChart historical={historicalCharacteristics} repositoryName={repo.name} />
                      )}
                    </Box>
                  </Collapse>
                </TableCell>
              </TableRow>
            </React.Fragment>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
}

export default RepositoriesTable;
