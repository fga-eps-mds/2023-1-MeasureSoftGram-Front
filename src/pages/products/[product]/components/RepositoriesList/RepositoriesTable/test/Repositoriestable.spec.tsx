import '@testing-library/jest-dom';

import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import { ProductProvider } from '@contexts/ProductProvider';
import RepositoriesTable from '../RepositoriesTable';

jest.mock('@contexts/RepositoryProvider', () => ({
  useRepositoryContext: () => ({
    repositoryList: [{ id: 19, name: '2022-1-MeasureSoftGram-Core' }]
  })
}));

jest.mock('@contexts/OrganizationProvider', () => ({
  useOrganizationContext: () => ({
    currentOrganization: { id: 1 }
  })
}));

const mockPush = jest.fn();

jest.mock('next/router', () => ({
  useRouter: () => ({
    push: mockPush
  })
}));

describe('RepositoriesTable', () => {
  describe('Functions', () => {
    test('chama a função handleRepositoriesFilter corretamente', () => {
      render(
        <ProductProvider>
          <RepositoriesTable />
        </ProductProvider>
      );
      const input = screen.getByRole('textbox', { name: /Insira o nome do repositório/i });
      fireEvent.change(input, { target: { value: 'Core' } });
      expect(screen.getByRole('table')).toHaveTextContent('2022-1-MeasureSoftGram-Core');
    });
  });
  describe('Snapshot', () => {
    it('Deve corresponder ao Snapshot', () => {
      const tree = render(
        <ProductProvider>
          <RepositoriesTable />
        </ProductProvider>
      );

      fireEvent.click(tree.getByTestId('repository-row'));

      expect(tree).toMatchSnapshot();
    });
  });
});
