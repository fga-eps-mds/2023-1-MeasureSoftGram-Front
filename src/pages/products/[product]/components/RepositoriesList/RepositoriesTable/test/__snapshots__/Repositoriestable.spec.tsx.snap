// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoriesTable Snapshot Deve corresponder ao Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
      >
        <table
          class="MuiTable-root css-rqglhn-MuiTable-root"
        >
          <thead
            class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                scope="col"
              >
                Nome
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                scope="col"
              />
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-hover css-1q1u3t4-MuiTableRow-root"
              data-testid="repository-row"
              style="cursor: pointer;"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              >
                2022-1-MeasureSoftGram-Core
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
              >
                <svg
                  aria-hidden="true"
                  aria-label="expand row"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ArrowCircleRightIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M22 12c0-5.52-4.48-10-10-10S2 6.48 2 12s4.48 10 10 10 10-4.48 10-10zm-10 1H8v-2h4V8l4 4-4 4v-3z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  aria-label="expand row"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="AddCircleIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
                  />
                </svg>
              </td>
            </tr>
            <tr
              class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                colspan="6"
                style="padding-bottom: 0px; padding-top: 0px;"
              />
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
    >
      <table
        class="MuiTable-root css-rqglhn-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Nome
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            />
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-hover css-1q1u3t4-MuiTableRow-root"
            data-testid="repository-row"
            style="cursor: pointer;"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              2022-1-MeasureSoftGram-Core
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
            >
              <svg
                aria-hidden="true"
                aria-label="expand row"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowCircleRightIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M22 12c0-5.52-4.48-10-10-10S2 6.48 2 12s4.48 10 10 10 10-4.48 10-10zm-10 1H8v-2h4V8l4 4-4 4v-3z"
                />
              </svg>
              <svg
                aria-hidden="true"
                aria-label="expand row"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="AddCircleIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
                />
              </svg>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              colspan="6"
              style="padding-bottom: 0px; padding-top: 0px;"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
